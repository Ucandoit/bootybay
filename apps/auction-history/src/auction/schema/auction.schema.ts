import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { Document } from 'mongoose';
export type AuctionDocument = Auction & Document;

@Schema({ collection: 'auctions' })
export class Auction {
  @Prop({ required: true })
  _id: string;

  @Prop({ required: true })
  itemId: number;

  @Prop({ required: true })
  realm: string;

  @Prop({ required: true })
  timestamp: number;

  @Prop()
  marketValue: number;

  @Prop()
  minBuyout: number;

  @Prop()
  historical: number;

  @Prop()
  numAuctions: number;

  @Prop()
  regionMarketValue: number;

  @Prop()
  regionHistorical: number;

  @Prop()
  regionSale: number;

  @Prop()
  regionSoldPerDay: number;

  @Prop()
  regionSalePercent: number;
}
export const AuctionSchema = SchemaFactory.createForClass(Auction);
