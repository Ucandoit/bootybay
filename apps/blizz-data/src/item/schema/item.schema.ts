import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { Document } from 'mongoose';
export type ItemDocument = Item & Document;

@Schema({ collection: 'items', strict: false })
export class Item {
  @Prop({ required: true, unique: true })
  id: number;

  @Prop({ required: true })
  name: string;
}
export const ItemSchema = SchemaFactory.createForClass(Item);
