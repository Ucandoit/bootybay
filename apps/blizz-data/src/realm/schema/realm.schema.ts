import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { Document } from 'mongoose';
export type RealmDocument = Realm & Document;

@Schema({ collection: 'realms' })
export class Realm {
  @Prop({ required: true, unique: true })
  id: number;

  @Prop({ required: true })
  slug: string;

  @Prop({ required: true })
  name: string;

  @Prop({ required: true })
  region: string;
}
export const RealmSchema = SchemaFactory.createForClass(Realm);
