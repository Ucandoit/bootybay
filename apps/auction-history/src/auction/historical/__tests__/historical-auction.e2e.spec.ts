import { INestApplication } from '@nestjs/common';
import { getConnectionToken, getModelToken, MongooseModule } from '@nestjs/mongoose';
import { Test } from '@nestjs/testing';
import { MongoMemoryServer } from 'mongodb-memory-server';
import { Connection, Model } from 'mongoose';
import * as request from 'supertest';
import { RealtimeAuctionRepository } from '../../realtime/realtime-auction.repository';
import {
  RealtimeAuction,
  RealtimeAuctionDocument,
  RealtimeAuctionSchema,
} from '../../realtime/realtime-auction.schema';
import { HistoricalAuctionController } from '../historical-auction.controller';
import { HistoricalAuctionRepository } from '../historical-auction.repository';
import { HistoricalAuction, HistoricalAuctionDocument, HistoricalAuctionSchema } from '../historical-auction.schema';
import { HistoricalAuctionService } from '../historical-auction.service';
import { historicalRegionalAuctionStub, realtimeRegionalAuctionStub } from './stubs/historical-auction.stub';

describe('Auction e2e tests', () => {
  let app: INestApplication;
  let memoryMongodb: MongoMemoryServer;
  let historicalAuctionModel: Model<HistoricalAuctionDocument>;
  let realtimeAuctionModel: Model<RealtimeAuctionDocument>;
  let connection: Connection;

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [
        MongooseModule.forFeature([
          { name: HistoricalAuction.name, schema: HistoricalAuctionSchema },
          { name: RealtimeAuction.name, schema: RealtimeAuctionSchema },
        ]),
        MongooseModule.forRootAsync({
          useFactory: async () => {
            memoryMongodb = await MongoMemoryServer.create();
            const mongoUri = await memoryMongodb.getUri();
            return {
              uri: mongoUri,
            };
          },
        }),
      ],
      controllers: [HistoricalAuctionController],
      providers: [HistoricalAuctionService, HistoricalAuctionRepository, RealtimeAuctionRepository],
    }).compile();

    app = moduleRef.createNestApplication();
    connection = moduleRef.get(getConnectionToken());
    historicalAuctionModel = moduleRef.get<Model<HistoricalAuctionDocument>>(getModelToken(HistoricalAuction.name));
    realtimeAuctionModel = moduleRef.get<Model<RealtimeAuctionDocument>>(getModelToken(RealtimeAuction.name));
    await app.init();
  });

  beforeEach(async () => {
    await historicalAuctionModel.deleteMany({});
    await realtimeAuctionModel.deleteMany({});
  });

  describe('POST /historical', () => {
    describe('request with invalid request body', () => {
      const testCases = [
        {
          description: 'should return 400 if body is not an array',
          body: {},
          status: 400,
        },
        {
          description: 'should return 400 if body is an empty array',
          body: [],
          status: 400,
        },
        {
          description: 'should return 400 if required itemId is missing',
          body: [
            {
              realm: 'bcc-eu-regional',
              timestamp: 1627580614000,
            },
          ],
          status: 400,
        },
        {
          description: 'should return 400 if required realm is missing',
          body: [
            {
              itemString: '1',
              timestamp: 1627580614000,
            },
          ],
          status: 400,
        },
        {
          description: 'should return 400 if required timestamp is missing',
          body: [
            {
              itemString: '1',
              realm: 'bcc-eu-regional',
            },
          ],
          status: 400,
        },
        {
          description: 'should return 400 if itemId is not a string',
          body: [
            {
              itemString: 1,
              realm: 'bcc-eu-regional',
              timestamp: 1627580614000,
            },
          ],
          status: 400,
        },
        {
          description: 'should return 400 if realm is not a string',
          body: [
            {
              itemString: '1',
              realm: 123,
              timestamp: 1627580614000,
            },
          ],
          status: 400,
        },
        {
          description: 'should return 400 if timestamp is not a number',
          body: [
            {
              itemString: '1',
              realm: 'bcc-eu-regional',
              timestamp: 'test',
            },
          ],
          status: 400,
        },
        {
          description: 'should return 400 if regionMarketValue is not a number',
          body: [
            {
              itemString: '1',
              realm: 'bcc-eu-regional',
              timestamp: 'test',
              regionMarketValue: 'test',
            },
          ],
          status: 400,
        },
        {
          description: 'should return 400 if regionHistorical is not a number',
          body: [
            {
              itemString: '1',
              realm: 'bcc-eu-regional',
              timestamp: 'test',
              regionHistorical: 'test',
            },
          ],
          status: 400,
        },
        {
          description: 'should return 400 if regionSale is not a number',
          body: [
            {
              itemString: '1',
              realm: 'bcc-eu-regional',
              timestamp: 'test',
              regionSale: 'test',
            },
          ],
          status: 400,
        },
        {
          description: 'should return 400 if regionSoldPerDay is not a number',
          body: [
            {
              itemString: '1',
              realm: 'bcc-eu-regional',
              timestamp: 'test',
              regionSoldPerDay: 'test',
            },
          ],
          status: 400,
        },
        {
          description: 'should return 400 if regionSalePercent is not a number',
          body: [
            {
              itemString: '1',
              realm: 'bcc-eu-regional',
              timestamp: 'test',
              regionSalePercent: 'test',
            },
          ],
          status: 400,
        },
        {
          description: 'should return 400 if marketValue is not a number',
          body: [
            {
              itemString: '1',
              realm: 'bcc-eu-regional',
              timestamp: 'test',
              marketValue: 'test',
            },
          ],
          status: 400,
        },
        {
          description: 'should return 400 if minBuyout is not a number',
          body: [
            {
              itemString: '1',
              realm: 'bcc-eu-regional',
              timestamp: 'test',
              minBuyout: 'test',
            },
          ],
          status: 400,
        },
        {
          description: 'should return 400 if historical is not a number',
          body: [
            {
              itemString: '1',
              realm: 'bcc-eu-regional',
              timestamp: 'test',
              historical: 'test',
            },
          ],
          status: 400,
        },
        {
          description: 'should return 400 if numAuctions is not a number',
          body: [
            {
              itemString: '1',
              realm: 'bcc-eu-regional',
              timestamp: 'test',
              numAuctions: 'test',
            },
          ],
          status: 400,
        },
      ];

      testCases.forEach(({ description, body, status }) => {
        test(description, async () => {
          const res = await request(app.getHttpServer()).post('/historical').send(body).expect(status);
          expect(res.body).toMatchSnapshot();
        });
      });
    });
  });

  describe('request with valid request body', () => {
    test('should return 201 with regional auction array', async () => {
      await request(app.getHttpServer()).post('/historical').send([historicalRegionalAuctionStub()]).expect(201);
      const historicalAuctionDocuments = await historicalAuctionModel.find({});
      expect(historicalAuctionDocuments).toHaveLength(1);
      expect(historicalAuctionDocuments[0]).toMatchSnapshot();
      const realtimeAuctionDocuments = await realtimeAuctionModel.find({});
      expect(realtimeAuctionDocuments).toHaveLength(1);
      expect(realtimeAuctionDocuments[0]).toMatchSnapshot();
    });

    test('should return 201 with realm auction array', async () => {
      await request(app.getHttpServer()).post('/historical').send([realtimeRegionalAuctionStub()]).expect(201);
      const historicalAuctionDocuments = await historicalAuctionModel.find({});
      expect(historicalAuctionDocuments).toHaveLength(1);
      expect(historicalAuctionDocuments[0]).toMatchSnapshot();
      const realtimeAuctionDocuments = await realtimeAuctionModel.find({});
      expect(realtimeAuctionDocuments).toHaveLength(1);
      expect(realtimeAuctionDocuments[0]).toMatchSnapshot();
    });

    test('unknown properties should be ignored', async () => {
      await request(app.getHttpServer())
        .post('/historical')
        .send([
          {
            ...historicalRegionalAuctionStub(),
            unknown: 'unknown',
          },
        ])
        .expect(201);
      const documents = await historicalAuctionModel.find({});
      expect(documents).toHaveLength(1);
      expect(documents[0]).toMatchSnapshot();
      expect(Reflect.hasOwnMetadata('unknown', documents[0])).toBe(false);
    });

    test('should return 500 if auction history already exists', async () => {
      await request(app.getHttpServer()).post('/historical').send([historicalRegionalAuctionStub()]).expect(201);
      await request(app.getHttpServer()).post('/historical').send([historicalRegionalAuctionStub()]).expect(500);
      // check that only one record is added
      const documents = await historicalAuctionModel.find({});
      expect(documents).toHaveLength(1);
      expect(documents[0]).toMatchSnapshot();
    });
  });

  afterAll(async () => {
    await connection.close();
    await memoryMongodb.stop();
    await app.close();
  });
});
