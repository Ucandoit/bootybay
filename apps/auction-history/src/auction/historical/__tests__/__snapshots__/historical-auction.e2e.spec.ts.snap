// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if body is an empty array 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"value\\" must contain at least 1 items",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if body is not an array 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"value\\" must be an array",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if historical is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"[0].timestamp\\" must be a number. \\"[0].historical\\" must be a number",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if itemId is not a string 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"[0].itemString\\" must be a string",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if marketValue is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"[0].timestamp\\" must be a number. \\"[0].marketValue\\" must be a number",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if minBuyout is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"[0].timestamp\\" must be a number. \\"[0].minBuyout\\" must be a number",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if numAuctions is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"[0].timestamp\\" must be a number. \\"[0].numAuctions\\" must be a number",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if realm is not a string 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"[0].realm\\" must be a string",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if regionHistorical is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"[0].timestamp\\" must be a number. \\"[0].regionHistorical\\" must be a number",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if regionMarketValue is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"[0].timestamp\\" must be a number. \\"[0].regionMarketValue\\" must be a number",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if regionSale is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"[0].timestamp\\" must be a number. \\"[0].regionSale\\" must be a number",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if regionSalePercent is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"[0].timestamp\\" must be a number. \\"[0].regionSalePercent\\" must be a number",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if regionSoldPerDay is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"[0].timestamp\\" must be a number. \\"[0].regionSoldPerDay\\" must be a number",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if required itemId is missing 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"[0].itemString\\" is required",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if required realm is missing 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"[0].realm\\" is required",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if required timestamp is missing 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"[0].timestamp\\" is required",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST /historical request with invalid request body should return 400 if timestamp is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"[0].timestamp\\" must be a number",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests request with valid request body should return 201 with realm auction array 1`] = `
Object {
  "__v": 0,
  "_id": "1-bcc-eu-mograine-horde-1627580614000",
  "historical": 1,
  "itemString": "1",
  "marketValue": 1.2,
  "minBuyout": 1.5,
  "numAuctions": 5,
  "realm": "bcc-eu-mograine-horde",
  "timestamp": 1627580614000,
}
`;

exports[`Auction e2e tests request with valid request body should return 201 with realm auction array 2`] = `
Object {
  "__v": 0,
  "_id": "1-bcc-eu-mograine-horde-1627580614000",
  "historical": 1,
  "itemString": "1",
  "marketValue": 1.2,
  "minBuyout": 1.5,
  "numAuctions": 5,
  "realm": "bcc-eu-mograine-horde",
  "timestamp": 1627580614000,
}
`;

exports[`Auction e2e tests request with valid request body should return 201 with regional auction array 1`] = `
Object {
  "__v": 0,
  "_id": "1-bcc-eu-regional-1627580614000",
  "itemString": "1",
  "realm": "bcc-eu-regional",
  "regionHistorical": 1.5,
  "regionMarketValue": 1.2,
  "regionSale": 1,
  "regionSalePercent": 0.2,
  "regionSoldPerDay": 5,
  "timestamp": 1627580614000,
}
`;

exports[`Auction e2e tests request with valid request body should return 201 with regional auction array 2`] = `
Object {
  "__v": 0,
  "_id": "1-bcc-eu-regional-1627580614000",
  "itemString": "1",
  "realm": "bcc-eu-regional",
  "regionHistorical": 1.5,
  "regionMarketValue": 1.2,
  "regionSale": 1,
  "regionSalePercent": 0.2,
  "regionSoldPerDay": 5,
  "timestamp": 1627580614000,
}
`;

exports[`Auction e2e tests request with valid request body should return 500 if auction history already exists 1`] = `
Object {
  "__v": 0,
  "_id": "1-bcc-eu-regional-1627580614000",
  "itemString": "1",
  "realm": "bcc-eu-regional",
  "regionHistorical": 1.5,
  "regionMarketValue": 1.2,
  "regionSale": 1,
  "regionSalePercent": 0.2,
  "regionSoldPerDay": 5,
  "timestamp": 1627580614000,
}
`;

exports[`Auction e2e tests request with valid request body unknown properties should be ignored 1`] = `
Object {
  "__v": 0,
  "_id": "1-bcc-eu-regional-1627580614000",
  "itemString": "1",
  "realm": "bcc-eu-regional",
  "regionHistorical": 1.5,
  "regionMarketValue": 1.2,
  "regionSale": 1,
  "regionSalePercent": 0.2,
  "regionSoldPerDay": 5,
  "timestamp": 1627580614000,
}
`;
